### PR TITLE
Implement the --django-debug plugin option with tests

### PR DESCRIPTION
**Pull request by dulacp**
Follows the #228 issue, and the un-merged #231 PR.

I've taken into account @blueyed [comment](https://github.com/pytest-dev/pytest-django/pull/231#issuecomment-93381011), and included a comprehensive list of tests.

Cheers

### CODE REVIEW

**Comment by karyon** [comment]
fwiw, starting from django 1.11 (to be released in april), django's test runner will have that option: https://docs.djangoproject.com/en/dev/ref/django-admin/#cmdoption-test-debug-mode

**Comment by blueyed** [comment]
@karyon 
Nice find.

It gets passed through to `setup_test_environment` (https://github.com/django/django/blob/5a6f70b4281817656db2f36c5919036d38fcce7f/django/test/runner.py#L461), which we should do then, too (in https://github.com/pytest-dev/pytest-django/blob/b3d7fd66416250f1c333dce49685e96bb89eabc4/pytest_django/plugin.py#L340).

@dulaccc 
Can you change this PR accordingly?
It should pass it in for Django 1.11+, but change the setting (as done currently) for older versions.

**Comment by dulaccc** [comment]
Of course I will update that in the afternoon!

**Comment by auvipy** [comment]
is this going out of radar?

**Comment by blueyed** [comment]
@dulaccc 
Can you update this please?  (we might have a release soon)

**Comment by dulacp** [comment]
I fixed the PR with the Django 1.11 new `setup_test_environment` API.
And I used the naming convention `--django-debug-mode`.

Is that okay for you?

Cheers

**Comment by karyon** [comment]
fyi, the tests failed :)

**Comment by dulacp** [comment]
@karyon are you sure the tests are stable between builds? because I've only 3 commits above the #4145f73 commit (which is marked as green on master). And the failed tests on travis are totally non-related to my additions.

I'll investigate,
Cheers

**Comment by karyon** [comment]
sorry, i'm not a maintainer of pytest-django, i just saw that the tests failed and thought i let you know because i myself often don't notice that :)

**Comment by dulacp** [comment]
Oh, alright. Thanks for the heads-up @karyon! :)

**Comment by dulacp** [comment]
Hey @blueyed, 

All the jobs targeting Python 2.7 are failing on Travis.
I suspect that part of the Travis stack changed during the past 15 days (because 15 days ago the commit 4145f73 was green).

Do you have other feedbacks on such an issue?

**Comment by blueyed** [comment]
Tests are failing with pytest 3.3 (and the other jobs are using an older version (cache likely?!))
See #550.

**Comment by blueyed** [comment]
Please rebase, since our tests in master were failing with pytest 3.3.

**Comment by dulacp** [comment]
@blueyed nice catch for the pytest version and the caching issue!
I guess everything is green now 🍻

**Comment by dulacp** [comment]
@blueyed can you merge this before master diverge?
Cheers

File: pytest_django/plugin.py [hunk 1]
```diff
@@ -66,6 +67,9 @@ def pytest_addoption(parser):
     group._addoption('--migrations',
                      action='store_false', dest='nomigrations', default=False,
                      help='Enable Django migrations on test setup')
+    group._addoption('--django-debug-mode',
+                     action='store', dest='djangodebugmode', default='False',
+                     help='Configure the DEBUG setting. Defaults to False')
```
Comment (blueyed) [inline pytest_django/plugin.py]: Shouldn't it default to `None`?
Also punctuation at the end?!

File: pytest_django/plugin.py [hunk 2]
```diff
@@ -332,11 +336,21 @@ def django_test_environment(request):
...
+        from distutils.version import StrictVersion
+        import django
         from django.conf import settings as dj_settings
         from django.test.utils import (setup_test_environment,
                                        teardown_test_environment)
-        dj_settings.DEBUG = False
-        setup_test_environment()
+        if request.config.getvalue('djangodebugmode') != 'None':
+            django_debug_mode = bool(strtobool(request.config.getvalue('djangodebugmode')))
+            if StrictVersion(django.get_version()) >= StrictVersion('1.11'):
```
Comment (blueyed) [inline pytest_django/plugin.py]: I think this should use the same mechanism for comparing versions like we do elsewhere, i.e. just via `django.VERSION`.

**Comment by samueldotj** [comment]
@dulacp  This issue took a solid 45 minutes to track it down - I guess a lot of other developers would have faced the same issue. It would be great if you could get this committed.

**Comment by richardARPANET** [comment]
Just wasted 30 minutes of my life due to this. Would be great to get this merged.

**Comment by blueyed** [comment]
@samueldotj @richardARPANET 
What exactly took you time to track it down?  IIRC this is about adding a new option to control the DEBUG setting only, isn't it?

Also it clearly cannot be "just merged" given 4 conflicting files.

**Comment by richardARPANET** [comment]
@blueyed My application was configured by env vars. One for the DEBUG setting.

```DEBUG = os.environ['DEBUG'].lower() == 'true'```

It worked at run-time, but not at test run-time. Which let to me trying/checking all sorts of things, I didn't expect pytest-django to be overriding this setting at all.

**Comment by blueyed** [comment]
I see.
FWIW this can also be controlled through settings, i.e.

```python
@pytest.fixture(scope="session", autouse=True)
def dj_debug(settings):
    settings.DEBUG = True
```

(or per test)

Django itself has `--debug-mode`, used by `setup_test_environment` (where we should/could pass this then also).

**Comment by blueyed** [comment]
@dulacp 
I'll pick it up later I guess - please let me know if you want / can do it instead.

@auvipy 
These kind of (obvious) "reviews" are not useful.


With regard to debugging you might be interested in https://github.com/pytest-dev/pytest-django/pull/725 - I'd appreciate a review of it in return.

File: pytest_django/plugin.py (hunk 1 above)
Comment (dulacp) [reply pytest_django/plugin.py]: Done.

File: pytest_django/plugin.py (hunk 2 above)
Comment (dulacp) [reply pytest_django/plugin.py]: Done.

**Comment by blueyed** [comment]
@dulacp 
Thanks!

I've pushed a fixup commit, let me know if this looks good to you, and we can finally (squash-)merge this then.

**Comment by dulacp** [comment]
Thanks for the fixup commit! It looks great :)

**Comment by blueyed** [comment]
Great.  Just wondered about the name - it mentions `--django-debug` initially, but is `django-debug-option`.  I'd prefer the shorter one, what do you think?
(sorry if this was changed / discussed before, just a quick thought)

**Comment by blueyed** [comment]
I think we should use `--django-debug` and `--django-no-debug` flags here.

**Comment by dulacp** [comment]
I've made the changes to use `django-debug` now.

Cheers

**Comment by Andrew-Chen-Wang** [comment]
@blueyed Should the master branch be merged to incorporate some new tests and pytest 5.4.0 or is this good to go?

**Comment by bluetech** [comment]
Oops, I was not aware of this PR and added a similar solution in #888. Only I used an ini option and called it `django_debug_mode`. Since that one is already merged, I'll close this one. Sorry for missing it, but let me know if you have any comments.

**Pull request closed without merging**